### ISSUE TITLE
Sort ranking seems bad for "json_rpc"

### ISSUE THREAD

**Issue by DanTup** [dart-lang/pub-dartlang-dart#1676]
I searched for "json_rpc" and got these results:

<img width="1029" alt="screen shot 2018-10-06 at 11 18 13" src="https://user-images.githubusercontent.com/1078012/46570274-8aa2d680-c959-11e8-92c6-a60fcbfb53a8.png">

The results are sorted by "relevance" which says:

> Packages are sorted by the combination of the text match, their overall score and their specificity to the selected platform. More information on ranking.

I'm not sure how they're weighted, but the top one doesn't seem like a great match (it mentions an API that also seems a bit off) and the second one is  *really* old with a score of 0 (even though it's a prefect name match). I don't know how easy it is to improve this, but I figured it may be worth raising for someone to review.

**Comment by isoos** [comment]
Thanks for the report!

The SDK result should be fixed once #1653 and other recently merge search PRs land.

The exact name match is tough. Would the order be still weird if the first SDK-API result would be gone, but the other two would remain the same order?

**Comment by DanTup** [comment]
I think it would a little, though I understand it may be hard to fix. I'd normally argue exact matches should always be at the top but this one is > 5 years old and has a score of 0 so the one underneath just feels so much more relevant.

I don't know, it's tough :) maybe packages below a certain threshold (of score/age/combination/whatever) should be collapsed/hidden by default with a note (similar to how Google search results have hide "duplicate" results with a link to include them anyway)?

(Out of interest, is the exact matched forced to the top because it's an exact match, or just ended up ranked highest based on some maths?)

**Comment by isoos** [comment]
> Out of interest, is the exact matched forced to the top because it's an exact match, or just ended up ranked highest based on some maths?

It is forced to the top because it is an exact match. Try searching the same string without the underscore, and the `_2` package will be at the top.

> maybe packages below a certain threshold (of score/age/combination/whatever)

I'm thinking of using the maintenance score as such threshold, because that will be automatically zero if the latest release is more than two years old. Exact match to the top should only happen if the maintenance score > 0.

**Comment by DanTup** [comment]
That sounds like a reasonable cut-off to me. It's slightly weird that things just "fall off a cliff", but I think two years is long enough to expect updates from a package expecting to be used (even if there are legitimately no changes, if a package is still being maintained it's not a big job to just bump the version).

**Comment by isoos** [comment]
Looks like #1677 was not enough, because the new package combiner also does a sort on merging results from two indexes, and it needs to have the same logic there.

**Comment by DanTup** [comment]
Not sure if it warrants another issue, but I also just noticed the third result for `xml` is `dart:convert`; but as far as I can tell there is absolutely no XML support in there (certainly it doesn't appear on any of the pages linked in the results):

<img width="979" alt="screenshot 2018-10-21 at 14 48 23" src="https://user-images.githubusercontent.com/1078012/47267797-67268100-d540-11e8-9fd4-42f7445fd804.png">

**Comment by isoos** [comment]
For the `xml` found in `dart:convert`, the core reason is this property's documentation:
https://api.dartlang.org/stable/2.0.0/dart-convert/AsciiCodec/name.html

The referenced URL is an `.xml` file, and the search index picked that up. Not yet sure whether and how we want to demote such results.

**Comment by DanTup** [comment]
Ah, not sure how I missed that! I guess that's somewhat valid then, though I think the ranking may still be off - there's an `xml2json` library that has `xml` in its name and a score of 92 appearing well below this one that only has a little mention in the description of a property.

**Comment by isoos** [comment]
Exact match results are fixed, tracking `xml` in https://github.com/dart-lang/pub-dartlang-dart/issues/1726.
